perf/lua/1mops_write_module.c

```diff
@@ -0,0 +1,173 @@
+#include <lua.h>
+#include <lauxlib.h>
+#include <module.h>
+#include <msgpuck.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "trivia/config.h"
+#include "trivia/util.h"
+
+/** Template tuples to update and insert: one per PK msgpack field size. */
+static char *base_tuple_pk1;
+static char *base_tuple_pk1_end;
+static char *base_tuple_pk2;
+static char *base_tuple_pk2_end;
+static char *base_tuple_pk3;
+static char *base_tuple_pk3_end;
+static char *base_tuple_pk5;
+static char *base_tuple_pk5_end;
+
+static bool
+find_base_tuple(uint32_t pk_value, char **tuple, char **tuple_end)
+{
+	size_t sizeof_pk_value = mp_sizeof_uint(pk_value);
+	switch (sizeof_pk_value) {
+	case 1:
+		*tuple = base_tuple_pk1;
+		*tuple_end = base_tuple_pk1_end;
+		return true;
+	case 2:
+		*tuple = base_tuple_pk2;
+		*tuple_end = base_tuple_pk2_end;
+		return true;
+	case 3:
+		*tuple = base_tuple_pk3;
+		*tuple_end = base_tuple_pk3_end;
+		return true;
+	case 5:
+		*tuple = base_tuple_pk5;
+		*tuple_end = base_tuple_pk5_end;
+		return true;
+	}
+	box_error_raise(ER_UNSUPPORTED, "No tuple for PK value of size %lu, "
+			"value: %u", sizeof_pk_value, pk_value);
+	return false;
+}
+
+static bool
+test_tuple(uint32_t pk_value, char **tuple, char **tuple_end)
+{
+	/* Get the tuple to update. */
+	if (!find_base_tuple(pk_value, tuple, tuple_end))
+		return false;
+
+	/* Check if the new PK value is of exact size. */
+	const char *data = *tuple;
+	mp_decode_array(&data);
+	char *pk_value_ptr = (char *)data;
+	uint32_t old_pk_value = mp_decode_uint(&data);
+	if (mp_sizeof_uint(pk_value) != mp_sizeof_uint(old_pk_value)) {
+		box_error_raise(ER_UNSUPPORTED, "Wrong base tuple, "
+				"PK value %u", pk_value);
+		return false;
+	}
+
+	/* Write the new PK value. */
+	mp_encode_uint(pk_value_ptr, pk_value);
+	return true;
+}
+
+static char *
+encode_base_tuple(char *data, ptrdiff_t *data_sz,
+		  uint32_t pk_value, uint32_t num_columns)
+{
+	data = mp_encode_array_safe(data, data_sz, num_columns);
+	data = mp_encode_uint_safe(data, data_sz, pk_value);
+	for (uint32_t i = 1; i < num_columns; i++)
+		data = mp_encode_uint_safe(data, data_sz, 0);
+	return data;
+}
+
+static bool
+create_base_tuples(uint32_t num_columns)
```

Comment:
I'm confused that we test engine by a limit number of datatypes. Should we parametrize a test (or probably add another one) that will test memcs by different datatypes (standard msgpack types and our extensions [^1]).


[^1]: https://www.tarantool.io/en/doc/latest/reference/internals/msgpack_extensions/